### PR TITLE
feat(al2023): cache and retag pause container

### PR DESCRIPTION
**Pull request by ndbaker1**
**Issue #, if available:**

**Description of changes:**

In the past we've added/removed pause container image caching due to various issues with image GC and regional images. here I'm bringing back image caching for AL2023 and retagging the image to remove the association with the image's registry

the new field `pause_container_image` will be necessary when building AMIs in an environment that cannot hit the configured default ecr uri (in `us-west-2`)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Testing Done**

*[See this guide for recommended testing for PRs.](../doc/CONTRIBUTING.md#testing-changes) Some tests may not apply. Completing tests and providing additional validation steps are not required, but it is recommended and may reduce review time and time to merge.*

### CODE REVIEW

**Comment by ndbaker1** [comment]
> Looks like we can probably use a shared provisioner for the caching logic?

yea good call, it might require a lot of the same env vars but i prefer the isolation & reuse with different provisioners

File: templates/al2/provisioners/install-worker.sh [hunk 1]
```diff
@@ -367,6 +367,21 @@ sudo chmod +x /etc/eks/bootstrap.sh
...
+# pull the pause container and cache it under localhost so that the regional
+# images do not matter from now on.
+
+sudo systemctl enable containerd --now
+PAUSE_CONTAINER="${ECR_URI}/eks/pause:${PAUSE_CONTAINER_VERSION}"
+sudo ctr --namespace k8s.io content fetch ${PAUSE_CONTAINER} --user AWS:$(aws ecr get-login-password --region $BINARY_BUCKET_REGION)
+sudo ctr --namespace k8s.io image tag ${PAUSE_CONTAINER} "localhost/kubernetes/pause:0.1.0"
+sudo ctr --namespace k8s.io image rm ${PAUSE_CONTAINER}
+# might not be necessary
+sudo ctr --namespace=k8s.io image label "localhost/kubernetes/pause:0.1.0" io.cri-containerd.pinned=pinned
```
[the remark this answers is not in the document]
Comment (ndbaker1) [reply templates/al2/provisioners/install-worker.sh]: yep, will do

File: templates/al2/provisioners/install-worker.sh [hunk 2]
```diff
@@ -367,6 +367,21 @@ sudo chmod +x /etc/eks/bootstrap.sh
...
 sudo chmod +x /etc/eks/max-pods-calculator.sh
 
+################################################################################
+### Pause Container ############################################################
+################################################################################
+
+# pull the pause container and cache it under localhost so that the regional
+# images do not matter from now on.
+
+sudo systemctl enable containerd --now
```
[the remark this answers is not in the document]
Comment (ndbaker1) [reply templates/al2/provisioners/install-worker.sh]: i was too broad here, updated to just start/stop in the scope of the provisioner

**Comment by ndbaker1** [comment]
gonna go ahead and sanity check this, i built both distros on local so it should be testable

/ci

File: templates/al2/template.json [hunk 3]
```diff
@@ -208,6 +208,18 @@
...
+    {
+      "type": "shell",
+      "remote_folder": "{{ user `remote_folder`}}",
+      "script": "{{template_dir}}/../shared/provisioners/cache-pause-container.sh",
+      "environment_vars": [
+        "AWS_ACCESS_KEY_ID={{user `aws_access_key_id`}}",
+        "AWS_SECRET_ACCESS_KEY={{user `aws_secret_access_key`}}",
+        "AWS_SESSION_TOKEN={{user `aws_session_token`}}",
+        "AWS_REGION={{user `aws_region`}}",
+        "PAUSE_CONTAINER_CMD=echo $(/etc/eks/get-ecr-uri.sh \"{{user `aws_region`}}\" \"$(imds 'latest/meta-data/services/domain')\")/eks/pause:3.5"
```
Comment (ndbaker1) [inline templates/al2/template.json:220]: i know what this looks like but hear me out 🤣 , the plumbing to get this script (which is also the map of region to accounts) is messy to sync between al2 and al23. i opted to just expose the mapping from `nodeadm` and allow each distro to pass in their expected handler code.

Comment (cartermckinnon) [reply templates/al2/template.json:220]: I do think this is kinda wonky, what if we moved the shared logic into a utility that could be used by a simple provisioner in each template?

Comment (ndbaker1) [reply templates/al2/template.json:220]: good with this, ill run another test sometime soon

**Comment by ndbaker1** [comment]
oh it only picked up al23 bc of the diff
/ci
+workflow:os_distros al2

**Comment by ndbaker1** [comment]
ah it wasn't cached for docker, which is still supported below 1.25 and default on 1.23 🤦

**Comment by ndbaker1** [comment]
/ci
+workflow:os_distros al2,al2023

**Comment by ndbaker1** [comment]
docker wasn't playing well with the importing the blob exported by `ctr`, so i just had to redo similar steps. not super critical since its a small image and 1.23/1.24 are soon to be EOL. should do the trick 🤞 

/ci
+workflow:os_distros al2,al2023

**Comment by ndbaker1** [comment]
rebased for conflicts on vendor modules

**Comment by ndbaker1** [comment]
retest after switching back to static image in `nodeadm`..
/ci

File: templates/al2023/provisioners/install-worker.sh
Comment (Issacwww) [inline templates/al2023/provisioners/install-worker.sh:1]: empty file??! is this been moved to somewhere else?

Comment (ndbaker1) [reply templates/al2023/provisioners/install-worker.sh:1]: 😂 it should still be here, must be a permission update i must have done based on the `100644 → 100755` but i didn't notice this message said empty. prob supposed to be "no diff"?

Comment (ndbaker1) [reply templates/al2023/provisioners/install-worker.sh:1]: ok fixed just for clarity 🥴

**Review by Issacwww — APPROVED** [review]

File: doc/usage/al2023.md [hunk 4]
```diff
@@ -67,3 +68,73 @@ One can pass the NVIDIA driver major version using the following:
 make k8s=1.29 os_distro=al2023 enable_accelerator=nvidia enable_efa=true nvidia_driver_major_version=560
 ```
 To see which driver versions are available, one can check the NVIDIA AL2023 [repository](https://developer.download.nvidia.com/compute/cuda/repos/amzn2023/).
+
+## Pause Container Image
+
+Since [PR #2000](https://github.com/awslabs/amazon-eks-ami/pull/2000), the pause
+container image is no longer pulled in at runtime on AL2023. `nodeadm` uses a
+static pause container reference to populate the
+[`sandbox`](https://github.com/containerd/containerd/blob/963c2160485631c3d4c1068aa28cea34c884a347/docs/cri/config.md?plain=1#L190)
```
Comment (cartermckinnon) [inline doc/usage/al2023.md:77]:
```suggestion
[`sandbox_image`](https://github.com/containerd/containerd/blob/963c2160485631c3d4c1068aa28cea34c884a347/docs/cri/config.md?plain=1#L190)
```

File: doc/usage/al2023.md [hunk 5]
```diff
@@ -67,3 +68,73 @@ One can pass the NVIDIA driver major version using the following:
...
+field of the containerd config.toml, which is baked into the AMI during the
+build based on the packer template variable `pause_container_image`.
+
+The default for `pause_container_image` is adequate for building in the `aws`
+partition, but if you require a region-specific ECR image, below is a table of
+the recommended pause container image for each region:
```
Comment (cartermckinnon) [inline doc/usage/al2023.md:83]: Could we get by with only one region per partition? It'd be nice to only update this table when a new partition comes online.

File: templates/shared/runtime/bin/cache-pause-container [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+while getopts 'i:t:e:' OPTION; do
+  case "$OPTION" in
+    i) PAUSE_CONTAINER_IMAGE="$OPTARG" ;;
+    t) TAG="$OPTARG" ;;
+    e) EXPORT_PATH="$OPTARG" ;;
+  esac
+done
+
+# pull the sandbox using aws credentials and tag it under a different name
+sudo ctr --namespace k8s.io image pull ${PAUSE_CONTAINER_IMAGE} --user AWS:$(aws ecr get-login-password)
```
Comment (cartermckinnon) [inline templates/shared/runtime/bin/cache-pause-container:19]: Try not to send our ECR creds to a non-ECR registry
```suggestion
PULL_ARGS=""
if [[ "${PAUSE_CONTAINER_IMAGE}" == *"dkr.ecr"* ]]; then
  PULL_ARGS="${PULL_ARGS} --user AWS:$(aws ecr get-login-password)"
fi
sudo ctr --namespace k8s.io image pull ${PAUSE_CONTAINER_IMAGE} ${PULL_ARGS}
```

File: doc/usage/al2023.md (hunk 4 above)
Comment (ndbaker1) [reply doc/usage/al2023.md:77]: whoops, i must have been too lazy and just grepped for sandbox 🫠, i fixed the underlying hyperlink as well
```suggestion
[`sandbox_image`](https://github.com/containerd/containerd/blob/963c2160485631c3d4c1068aa28cea34c884a347/docs/cri/config.md?plain=1#L314-L315)
```

File: doc/usage/al2023.md (hunk 5 above)
Comment (ndbaker1) [reply doc/usage/al2023.md:83]: true, ill double check if that'll work

Comment (ndbaker1) [reply doc/usage/al2023.md:83]: i think the opt-in regions wont play nice with this, but i had another idea. what if we just reuse the EKS addon docs for this table?

https://docs.aws.amazon.com/eks/latest/userguide/add-ons-images.html

**Comment by cartermckinnon** [comment]
Sanity check before merge...

/ci

**Comment by cartermckinnon** [comment]
Argh sorry @ndbaker1 😭  Can you rebase this when you get a chance?

**Pull request merged**